[14.0] shopfloor: Cluster picking / Location content transfer - Fix error messages in case of wrong thing scanned

jbaudoux opened the pull request:
If you don't scan the right package/product/lot/location, raise the correct error to the operator that he scanned the wrong package/product/lot/location instead of raising a "Barcode not found"

In location content transfer, initial step where you have to scan a location:
* add a check that the location is valid for the menu's picking types
* raise a "Location empty" error message if the location is empty instead of a "no pack found in location" (the pack error is for single_pack_transfer)

I changed the "Location empty" from info to error. I checked the scenario using this message and seems to me to always be an error and not an info.

cc @sebalix @simahawk @lmignon @Lmarion-source

jbaudoux committed:
shopfloor: Cluster picking - Fix wrong thing scanned

If you don't scan the right package/product/lot/location, raise the
correct error to the operator that he scanned the wrong
package/product/lot/location instead of raising a "Barcode not found"

simahawk commented on shopfloor/actions/message.py:
```diff
@@ -199,12 +199,24 @@ def no_product_for_barcode(self, barcode):
             "body": _("No product found for {}").format(barcode),
         }
 
+    def wrong_package(self):
+        return {
+            "message_type": "error",
+            "body": _("Wrong pack."),
+        }
+
```
```suggestion
    def _wrong_record_msg(self, model_name):
        return {
            "product.product": _("Wrong product."),
            "stock.location": _("Wrong location."),
            "stock.quant.package": _("Wrong pack."),
        }.get(model_name, _("Wrong record"))
        
    def wrong_record(self, record):
        return {
            "message_type": "error",
            "body": self._wrong_record_msg(record._name)
        }

```

jbaudoux replied:
:+1:

simahawk commented on shopfloor/actions/message.py:
```diff
@@ -199,12 +199,24 @@ def no_product_for_barcode(self, barcode):
             "body": _("Wrong product."),
         }
 
+    def wrong_location(self):
+        return {
+            "message_type": "error",
+            "body": _("Wrong location."),
+        }
+
```
```suggestion
```

simahawk commented on shopfloor/services/cluster_picking.py:
```diff
@@ -464,6 +464,23 @@ def scan_line(self, picking_batch_id, move_line_id, barcode):
             return self._scan_line_by_location(picking, move_line, location)
 
+        # Nothing matches what is expected from the move line.
+        if package:
+            return self._response_for_start_line(
+                move_line, message=self.msg_store.wrong_package()
+            )
+        if product:
+            return self._response_for_start_line(
+                move_line, message=self.msg_store.wrong_product()
+            )
+        if lot:
+            return self._response_for_start_line(
+                move_line, message=self.msg_store.wrong_lot()
+            )
+        if location:
+            return self._response_for_start_line(
+                move_line, message=self.msg_store.wrong_location()
+            )
```
```suggestion
        for rec in (package, product, lot, location):
            if rec:
                return self._response_for_start_line(
                    move_line, message=self.msg_store.wrong_record(rec)
                )
```

simahawk commented on shopfloor/actions/message.py:
```diff
@@ -199,11 +199,16 @@ def no_product_for_barcode(self, barcode):
             "body": _("No product found for {}").format(barcode),
         }
 
-    def wrong_product(self):
```
this way you break https://github.com/OCA/wms/blob/b84a3ef9d4d3f9667707316a1fb7afa76a8dc796/shopfloor_manual_product_transfer/services/manual_product_transfer.py#L483

Please:
1. do not remove this method
2. make it use the new one
3. log a deprecation warning if we expect to drop it completely for v15
4. adapt shopfloor_manual_product_transfer to use the new method

jbaudoux replied:
Thanks, done

simahawk reviewed (changes requested):
@jbaudoux any reason to keep this in draft?

jbaudoux committed:
shopfloor: Location content transfer - Fix wrong thing scanned

If you don't scan the right package/product/lot, raise the
correct error to the operator that he scanned the wrong
package/product/lot instead of raising a "Barcode not found"

In the initial step where you have to scan a location:
* add a check that the location is valid for the menu's picking types
* raise a "Location empty" error message if the location is empty
  instead of a "no pack found in location" (the pack error is for
  single_pack_transfer)

I changed the "Location empty" from info to error. I checked the
scenario using this message and seems to me to always be an error and
not an info.

sebalix commented on shopfloor/actions/message.py:
```diff
@@ -200,11 +204,24 @@ def no_product_for_barcode(self, barcode):
+            "stock.production.lot": _("Wrong lot."),
+            "stock.location": _("Wrong location."),
+            "stock.quant.package": _("Wrong pack."),
+        }.get(model_name, _("Wrong."))
```
![](https://media4.giphy.com/media/3o6Mb43PiNTQS5WgLu/giphy.gif)

simahawk commented on shopfloor/actions/message.py:
```diff
@@ -218,9 +235,11 @@ def lot_on_wrong_product(self, barcode):
 
     def wrong_lot(self):
+        # Method to drop in v15
+        _logger.warning("`wrong_product` is deprecated, use `wrong_record` instead")
```
```suggestion
        _logger.warning("`wrong_lot` is deprecated, use `wrong_record` instead")
```

jbaudoux replied:
:+1:

simahawk commented:
/ocabot merge patch
